agimus_controller_ros/agimus_controller_ros/agimus_controller.py

```diff
@@ -0,0 +1,250 @@
+        self.rmodel = pin.buildReducedModel(
+            self.rmodel,
+            list_of_geom_models=[],
+            list_of_joints_to_lock=locked_joint_ids,
+            reference_configuration=np.zeros(self.rmodel.nq),
+        )[0]
+        self.rdata = self.rmodel.createData()
+        self.effector_frame_id = self.rmodel.getFrameId(
+            self.params.ocp.effector_frame_name
+        )
+        self.nq = self.rmodel.nq
+        self.nv = self.rmodel.nv
+        self.nx = self.nq + self.nv
+        self.get_logger().info("robot_description_callback")
+        self.get_logger().info(f"pin_model.nq {self.rmodel.nq}")
+
+    def wait_first_sensor_msg(self):
+        wait_for_input = True
+        while not rclpy.is_shutdown() and wait_for_input:
+            wait_for_input = (
+                not self.first_robot_sensor_msg_received
+                or not self.first_pose_ref_msg_received
+            )
+            if wait_for_input:
+                rclpy.get_logger().info(3, "Waiting until we receive a sensor message.")
+            rclpy.get_logger().info("Start controller")
+            self.rate.sleep()
+        return wait_for_input
+
+    def wait_buffer_has_twice_horizon_points(self):
+        while (
+            self.traj_buffer.get_size(self.point_attributes)
+            < 2 * self.params.ocp.horizon_size
+        ):
+            self.fill_buffer()
+
+    def get_next_trajectory_point(self):
+        raise RuntimeError("Not implemented")
+
+    def fill_buffer(self):
+        point = self.get_next_trajectory_point()
+        if point is not None:
+            self.traj_buffer.add_trajectory_point(point)
+        while point is not None:
+            point = self.get_next_trajectory_point()
+            if point is not None:
+                self.traj_buffer.add_trajectory_point(point)
+
+    def first_solve(self, x0):
+        # retrieve horizon state and acc references
+        horizon_points = self.traj_buffer.get_points(
+            self.params.ocp.horizon_size, self.point_attributes
+        )
+        x_plan = np.zeros([self.params.ocp.horizon_size, self.nx])
+        a_plan = np.zeros([self.params.ocp.horizon_size, self.nv])
+        for idx_point, point in enumerate(horizon_points):
+            x_plan[idx_point, :] = point.get_x_as_q_v()
+            a_plan[idx_point, :] = point.a
+
+        # First solve
+        self.ocp = OCPCrocoHPP(self.rmodel, self.cmodel, self.params.ocp)
+        self.ocp.set_planning_variables(x_plan, a_plan)
+        self.mpc = MPC(self.ocp, x_plan, a_plan, self.rmodel, self.cmodel)
+        us_init = self.mpc.ocp.u_plan[: self.params.ocp.horizon_size - 1]
+        self.mpc.mpc_first_step(x_plan, us_init, x0)
+        self.next_node_idx = self.params.ocp.horizon_size
+        if self.params.save_predictions_and_refs:
+            self.mpc.create_mpc_data()
+        _, u, k = self.mpc.get_mpc_output()
+        return u, k
```

Comment:
Shouldn't aprt of it happen inside if `__init__`, part when we know URDF and part of this isn't expected to be handled in `MPC` class itself?

Reply:
in the `__init__ ` we create the callback to retrieve URDF but we have to wait for it, some of the steps here may change with the new architecture but for now with the old architecture of mpc/ocp class it's done this way